README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![](https://img.shields.io/badge/lifecycle-mature-blue.svg)](https://www.tidyverse.org/lifecycle/#mature)
[![](https://img.shields.io/github/last-commit/mikkelkrogsholm/statsDK.svg)](https://github.com/mikkelkrogsholm/statsDK/commits/master)
[![](https://img.shields.io/badge/devel%20version-0.1.1-blue.svg)](https://github.com/mikkelkrogsholm/statsDK)

[![](https://www.r-pkg.org/badges/version/statsDK?color=orange)](https://cran.r-project.org/package=statsDK)
[![](http://cranlogs.r-pkg.org/badges/grand-total/badger?color=orange)](https://cran.r-project.org/package=badger)

# statsDK

The goal of statsDK is to make it easy to call the API of Statistics
Denmark.

## Installation

You can install statsDK from github with:

``` r
# install.packages("devtools")
devtools::install_github("mikkelkrogsholm/statsDK")
```

## Example

This little vignette shows you how to get started with the `statsDK`
package.

## The retrievers

The package has a few “retriever”-functions that are used to retrieve
data from Statistics Denmark. Those are: `sdk_retrieve_subjects()`,
`sdk_retrieve_tables()`, `sdk_retrieve_metadata()` and the
`sdk_retrieve_data()` functions.

### retrieve\_subjects

This function retrieves the overall subjects that are available in the
API.

### retrieve\_tables

This function retrieves an overview of all the tables in the API. Lets
use it to see what data we would like to
fetch:

``` r
library(statsDK); library(dplyr); library(stringr); library(lubridate); library(ggplot2); library(tidyr)

tables <- sdk_retrieve_tables()

glimpse(tables)
#> Observations: 2,051
#> Variables: 8
#> $ id           <chr> "FOLK1A", "FOLK1B", "FOLK1C", "FOLK1D", "FOLK1E", "…
#> $ text         <chr> "Population at the first day of the quarter", "Popu…
#> $ unit         <chr> "number", "number", "number", "number", "number", "…
#> $ updated      <chr> "2019-08-09T08:00:00", "2019-08-09T08:00:00", "2019…
#> $ firstPeriod  <chr> "2008Q1", "2008Q1", "2008Q1", "2008Q1", "2008Q1", "…
#> $ latestPeriod <chr> "2019Q3", "2019Q3", "2019Q3", "2019Q3", "2019Q3", "…
#> $ active       <lgl> TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRU…
#> $ variables    <list> [<"region", "sex", "age", "marital status", "time"…
```

Lets say we are interested in marriages. Maybe there is an official data
set about marriages that we can use?

First we unnest the variables column:

``` r
tables_long <- tables %>%
  unnest(variables)
```

Then we see if we can find something with marriage. We use the
`str_detect()` function from the `stringr` package to detect a text
pattern matching marriage:

``` r

marriage_tables <- tables_long %>%
  filter(str_detect(tables_long$variables, "marriage"))

glimpse(marriage_tables)
#> Observations: 10
#> Variables: 8
#> $ id           <chr> "VIEDAG", "VIEDAG", "VIE8", "VIE6", "VIE307", "VIE3…
#> $ text         <chr> "Marriages", "Marriages", "Marriages between two of…
#> $ unit         <chr> "number", "number", "number", "number", "number", "…
#> $ updated      <chr> "2019-02-14T08:00:00", "2019-02-14T08:00:00", "2019…
#> $ firstPeriod  <chr> "2007", "2007", "2007", "2012", "2006", "2006", "20…
#> $ latestPeriod <chr> "2018", "2018", "2018", "2018", "2018", "2018", "20…
#> $ active       <lgl> TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRU…
#> $ variables    <chr> "day of marriage", "month of the marriage", "type o…
```

Indeed there is. There is the `VIEDAG` table that seems to have data on
marriage that we might be interested in. Lets therefore have a look at
the meta data for that particular table.

### retrieve\_metadata

This function retrieves meta data for a table - like our `VIEDAG` table.

``` r
viedag_meta <- sdk_retrieve_metadata("VIEDAG")
#> Metadata collected succesfully

glimpse(viedag_meta)
#> List of 11
#>  $ id                 : chr "VIEDAG"
#>  $ text               : chr "Marriages"
#>  $ description        : chr "Marriages by day of marriage, month of the marriage and time"
#>  $ unit               : chr "number"
#>  $ suppressedDataValue: chr "0"
#>  $ updated            : chr "2019-02-14T08:00:00"
#>  $ active             : logi TRUE
#>  $ contacts           :'data.frame': 1 obs. of  3 variables:
#>   ..$ name : chr "Connie Østberg"
#>   ..$ phone: chr "+4539173384"
#>   ..$ mail : chr "cbn@dst.dk"
#>  $ documentation      :List of 2
#>   ..$ id : chr "fc512ffc-7334-4237-aab9-b776fcc6748c"
#>   ..$ url: chr "https://www.dst.dk/documentationofstatistics/fc512ffc-7334-4237-aab9-b776fcc6748c"
#>  $ footnote           : NULL
#>  $ variables          :'data.frame': 3 obs. of  5 variables:
#>   ..$ id         : chr [1:3] "VDAG" "VIMDR" "Tid"
#>   ..$ text       : chr [1:3] "day of marriage" "month of the marriage" "time"
#>   ..$ elimination: logi [1:3] TRUE TRUE FALSE
#>   ..$ time       : logi [1:3] FALSE FALSE TRUE
#>   ..$ values     :List of 3
#>   .. ..$ :'data.frame':  32 obs. of  2 variables:
#>   .. ..$ :'data.frame':  13 obs. of  2 variables:
#>   .. ..$ :'data.frame':  12 obs. of  2 variables:
```

The list of meta data has a lot of information that we can use to
determine wether or not to use the data. There is an URL under
documentation that we can follow to read a lot more about the data and
how it is collected. There is contact information if we still have
unanswered questions that need to be answered.

And there is also a part of the list called variables. This is the part
we need to determine what we can get from calling that table directly
and also how we should call it. We will use the helper function
`get_variables()` to get a nice tidy tibble to inspect.

``` r
variables <- sdk_get_variables(viedag_meta)

glimpse(variables)
#> Observations: 57
#> Variables: 4
#> $ param       <chr> "VDAG", "VDAG", "VDAG", "VDAG", "VDAG", "VDAG", "VDA…
#> $ setting     <chr> "TOT", "D01", "D02", "D03", "D04", "D05", "D06", "D0…
#> $ type        <chr> "day of marriage", "day of marriage", "day of marria…
#> $ description <chr> "Total", "1.", "2.", "3.", "4.", "5.", "6.", "7.", "…
```

Lets see if we can get a short overview of all the different options we
have. Lets make a tibble for this vignette that shows the first, middle
and last row of each parameter:

``` r
variable_overview <- variables %>% 
  group_by(param) %>%
  slice(c(1, round(n()/2), n())) %>%
  ungroup()

variable_overview
#> # A tibble: 9 x 4
#>   param setting type                  description
#>   <chr> <chr>   <chr>                 <chr>      
#> 1 Tid   2007    time                  2007       
#> 2 Tid   2012    time                  2012       
#> 3 Tid   2018    time                  2018       
#> 4 VDAG  TOT     day of marriage       Total      
#> 5 VDAG  D15     day of marriage       15.        
#> 6 VDAG  D31     day of marriage       31.        
#> 7 VIMDR TOT     month of the marriage Total      
#> 8 VIMDR 005     month of the marriage May        
#> 9 VIMDR 012     month of the marriage December
```

From this overview it looks like `Tid` is the year, `VDAG` is the day of
the month and `VIMDR` is the month. `VDAG` and `VIMDR` also has a `TOT`
that is the total.

With this newfound knowledge we can now construct an API call to get the
data we are interested in.

### retrieve\_data

This is the function that actually retrieves the data that we need.

Lets get the total data for each month of june and december for all the
available years. This forces us to construct an API call that shows
different aspects.

From the variable overview we did earlier we can see that in order to
get the `Total` for days of marriage then we have to use the `TOT`
setting for the `VDAG` parameter. And in order to get the month of
`June` and `December` we will have to use the `006` and `012` setting
for the `VIMDR` parameter. But how do we call all years? Easy, we just
have to set that to be an asterix `*`.

Below is the call to the
API:

``` r
VIEDAG <- sdk_retrieve_data("VIEDAG", Tid = "*", VDAG = "TOT", VIMDR = "006,012")
#> Getting data. This can take a while, if the data is very large.
#> Data collected succesfully
names(VIEDAG) <- c("time", "day", "month", "count")
```

Let us have a glimpse at the data:

``` r
glimpse(VIEDAG)
#> Observations: 24
#> Variables: 4
#> $ time  <dbl> 2007, 2007, 2008, 2008, 2009, 2009, 2010, 2010, 2011, 2011…
#> $ day   <chr> "Total", "Total", "Total", "Total", "Total", "Total", "Tot…
#> $ month <chr> "June", "December", "June", "December", "June", "December"…
#> $ count <dbl> 4486, 2092, 3838, 1894, 3546, 1824, 3396, 1596, 3462, 1394…
```

Finally lets plot it and see what is going on in our new marriage data
set:

``` r

VIEDAG$time <- ymd(paste0(VIEDAG$time, "-01-01"))

my_y <- VIEDAG %>%
  filter(time == max(time)) %>%
  pull(count)

ggplot(VIEDAG) +
  geom_line(aes(x = time, count, group = month)) +
  annotate("text", x = max(VIEDAG$time) %m+% months(1) , y = my_y, 
           label = c("June", "December"), hjust = 0) +
  annotate("point", x = max(VIEDAG$time), y = my_y) +
  xlim(min(VIEDAG$time), max(VIEDAG$time) %m+% years(1) ) +
  labs(y = "Total marriages for the given month", x = "Years") +
  theme_minimal()
```

![](README-unnamed-chunk-10-1.png)<!-- -->

There is quite a spike in the data for December 2012. A lot of people
got married in December in that particular year…

Can you figure out why? Make your own API call that calls all days in
December for all years and see if you can figure out what made that
particular year so different…

## Further ressources

Visit the <http://statbank.dk/> and <http://api.statbank.dk/console> for
further exploration of Statistics Denmark data.


NEWS.md
# statsDK 0.1.0

* This was the first version

# statsDK 0.1.1
* Added a `NEWS.md` file to track changes to the package.
* Changed the package to use the streaming version of the API. This ensures it can collect the very large data sets.
* Changed to using the httr package for retrival in order to ensure proper error messages from the API.
* Added the base_url and language arguments for further flexibility.
* retrieve_data() will fill out missing parameters and setting them to *, which means all data.
* Metada for a table is added to it as an attribute.
* Changed the fix_time() function to work on strings and not data frames. The reason for this was to increase the flexibility in cause of name changes in column names.

# statsDK 1.0.0
* Breaking changes. changed function names to all begin with `sdk_` to adhere to newest pacakge building best practices and to avoid potential namespace polution.
* Fixed an error resulting from an update to the `tidyr` package that got it removed from CRAN.


tests/testthat/test_fixers.R
library(statsDK)
library(stringr)
library(testthat)

# fix_time ----
context("Testing fix_time")

folk_1 <- suppressMessages(
  sdk_retrieve_data("FOLK1A", TID = "*", ALDER = "IALT", CIVILSTAND = "TOT", lang = "da")
)

folk_2 <- folk_1

folk_2$TID <- sdk_fix_time(folk_2$TID)

test_that("the dates are as expected", {
  expect_true(all(stringr::str_detect(folk_1$TID, "\\d{4}\\w\\d{1}")))
  expect_true(all(stringr::str_detect(folk_2$TID, "\\d{4}-\\d{2}-\\d{2}")))
})

# get_variables ----
context("Testing get_variables")

metadata <- sdk_retrieve_metadata("BEV3A")

# See the variables as a data frame
variables <- sdk_get_variables(metadata)

test_that("we have a nice tidy data frame", {
  expect_true(is.data.frame(variables))
  expect_true(nrow(variables) > 0)
  expect_true(ncol(variables) > 0)
})


tests/testthat/test_retrievers.R
library(statsDK)
library(testthat)

# retrieve_subjects ----
context("Testing retrieve_subjects")

subjects <- sdk_retrieve_subjects()

test_that("we have indeed received data for subjects", {
  expect_true(is.data.frame(subjects))
  expect_true(nrow(subjects) > 0)
  expect_true(ncol(subjects) > 0)
})

# retrieve_tables ----
context("Testing retrieve_tables")

tables <- sdk_retrieve_tables()

test_that("we have indeed received data for tables", {
  expect_true(is.data.frame(tables))
  expect_true(nrow(tables) > 0)
  expect_true(ncol(tables) > 0)
})


# retrieve_metadata ----
context("Testing retrieve_metadata")

folk_meta <- sdk_retrieve_metadata("FOLK1A")

test_that("we have indeed received meta data", {
  expect_true(is.list(folk_meta))
  expect_true(length(folk_meta) > 0)
  expect_true(folk_meta$id == "FOLK1A")
})

# retrieve_data ----
context("Testing retrieve_data")

pris <- suppressMessages(
  sdk_retrieve_data("PRIS111", VAREGR = "000000", ENHED = "*",
                Tid = paste(paste0("2017M0", 1:8), collapse = ","))
  )

test_that("we have indeed received data", {
  expect_true(is.data.frame(pris))
  expect_true(nrow(pris) > 0)
  expect_true(ncol(pris) > 0)
})




R/fixer_functions.R
# fix_time ----

#' Fixes the time column of a data set
#'
#' Statistics Denmark data often come with times/dates that are formatted to
#' fit months or quarters. This function converts them in to proper date strings
#' in the format YYYY-MM-DD.
#'
#' @param date_string a string of dates formatted as months or quarters.
#' @param as_char whether to return the dates as data objects or character strings.
#'
#' @return a data frame
#' @export
#'
#' @examples
#' df <- statsDK::sdk_retrieve_data("FOLK1A", TID = "*", ALDER = "IALT",
#'                              CIVILSTAND = "TOT", lang = "da")
#' dplyr::glimpse(df)
#'
#' df$TID <- statsDK::sdk_fix_time(df$TID)
#' dplyr::glimpse(df)
#'
sdk_fix_time <- function(date_string, as_char = FALSE){

  # Fix quarters
  date_string <- gsub('(\\d{4})(Q|K)(1)', '\\1-01-01', date_string)
  date_string <- gsub('(\\d{4})(Q|K)(2)', '\\1-04-01', date_string)
  date_string <- gsub('(\\d{4})(Q|K)(3)', '\\1-07-01', date_string)
  date_string <- gsub('(\\d{4})(Q|K)(4)', '\\1-10-01', date_string)

  # Fix months
  date_string <- gsub('(\\d{4})(M)(\\d{2})', '\\1-\\3-01', date_string)

  # Convert to data format
  if(!as_char) date_string <- lubridate::ymd(date_string)

  # Return df with fixed time
  return(date_string)
}

# get_variables ----

#' Extracts variables from the list of meta data
#'
#' Extracts the variables and sets them up in a tibble so it is easy to see
#' what settings each parameter should have to get the desired data.
#'
#' @param metadata A list retrieved using the \link{sdk_retrieve_metadata} function.
#'
#' @return a tibble
#' @export
#'
#' @examples
#' metadata <- statsDK::sdk_retrieve_metadata("BEV3A")
#' dplyr::glimpse(metadata)
#'
#' # See the variables as a data frame
#' variables <- sdk_get_variables(metadata)
#' dplyr::glimpse(variables)
#'
sdk_get_variables <- function(metadata){

  variables_df <- tibble::as_tibble(metadata$variables)

  variables <- tidyr::unnest(variables_df, cols = "values", names_repair = "minimal")

  names(variables) <- c("param", "type", "x", "y", "setting", "description")

  variables <- variables[, c("param", "setting", "type", "description")]

  variables <- tibble::as_tibble(variables)

  return(variables)

}


R/retrieve_functions.R
# retrieve_subjects ----

#' Retrieves an overview of the subjects from Statistics Denmark
#'
#' The function retrieves a data frame with the over all statistical subjects
#' that can be pulled from Statistics Denmark.
#'
#' @param base_url is the base url for the API you wish to call. Statistics Denmark
#'     can sometimes create custom API's that you can use by changing this
#'     parameter.
#'
#' @return a tibble
#' @export
#'
#' @examples
#' subjects <- statsDK::sdk_retrieve_subjects()
#' dplyr::glimpse(subjects)

sdk_retrieve_subjects <- function(base_url = "http://api.statbank.dk/v1/"){

  url <- paste0(base_url, "subjects?lang=en&format=JSON")

  # Get data ----
  get_data <- suppressWarnings(httr::GET(url))

  # Decide if error ----
  if(get_data$status_code != 200){
    content_data <- httr::content(get_data, "text", encoding = "UTF-8")
    content_data <- jsonlite::fromJSON(content_data)
    my_message <- paste(content_data$message, collapse = "\n")
    message(my_message)
    return(NULL)
  } else {
    message("Subjects collected succesfully")
  }

  content_data <- httr::content(get_data, "text", encoding = "UTF-8")
  content_data <- jsonlite::fromJSON(content_data)

  content_data$subjects <- NULL

  content_data <- tibble::as_tibble(content_data)

  return(content_data)
}

# retrieve_tables ----

#' Retrieves an overview of the tables from Statistics Denmark
#'
#' The function retrieves a data frame with all the available tables from the
#' Statistics Denmark API.
#'
#' \describe{
#'   \item{id}{The id of the table. This is used when calling specific tables
#'       later with \link{sdk_retrieve_metadata} or \link{sdk_retrieve_data}.}
#'   \item{text}{A description of what the data in the table is about.}
#'   \item{unit}{What unit the data is in.}
#'   \item{updated}{When the table was last updated}
#'   \item{firstPeriod}{The first period in the data, ie how far back the data goes.}
#'   \item{latestPeriod}{The latest period in the data.}
#'   \item{active}{If the table is still being updated}
#'   \item{variables}{A list of the variables in the tables.}
#' }
#'
#' @param base_url is the base url for the API you wish to call. Statistics Denmark
#'     can sometimes create custom API's that you can use by changing this
#'     parameter.
#'
#' @return a tibble
#' @export
#'
#' @examples
#' tables <- statsDK::sdk_retrieve_tables()
#' dplyr::glimpse(tables)

sdk_retrieve_tables <- function(base_url = "http://api.statbank.dk/v1/"){

  url <- paste0(base_url, "tables?lang=en&format=JSON")

  # Get data ----
  get_data <- suppressWarnings(httr::GET(url))

  # Decide if error ----
  if(get_data$status_code != 200){
    content_data <- httr::content(get_data, "text", encoding = "UTF-8")
    content_data <- jsonlite::fromJSON(content_data)
    my_message <- paste(content_data$message, collapse = "\n")
    message(my_message)
    return(NULL)
  } else {
    message("Tables collected succesfully")
  }

  content_data <- httr::content(get_data, "text", encoding = "UTF-8")

  content_data <- jsonlite::fromJSON(content_data)

  content_data <- tibble::as_tibble(content_data)

  return(content_data)
}

# retrieve_metadata ----

#' Retrieves metadata for a specific table from Statistics Denmark
#'
#' The function retrieves a list with a lot of metadata about a certain table.
#'
#' \describe{
#'   \item{id}{The id of the table.}
#'   \item{text}{A description of what the data in the table is about.}
#'   \item{unit}{What unit the data is in.}
#'   \item{contact}{Who to contact regarding the data.}
#'   \item{documentation}{Link to a web page with detailed description of the data.}
#'   \item{footnote}{A footnote if applicable.}
#'   \item{variables}{Details about the variables in the table. Is very useful for
#'       when using the \link{sdk_retrieve_data} function.}
#' }
#'
#' @param table_id is the id of the table you want to call. You can get table ids
#'     by calling the \link{sdk_retrieve_tables} function.
#' @param base_url is the base url for the API you wish to call. Statistics Denmark
#'     can sometimes create custom API's that you can use by changing this
#'     parameter.
#'
#' @return a list
#' @export
#'
#' @examples
#' metadata <- statsDK::sdk_retrieve_metadata("PRIS111")
#' dplyr::glimpse(metadata)

sdk_retrieve_metadata <- function(table_id, base_url = "http://api.statbank.dk/v1/"){

  url <- paste0(base_url, "tableinfo/",
                table_id, "?lang=en&format=JSON")

  # Get data ----
  get_data <- suppressWarnings(httr::GET(url))

  # Decide if error ----
  if(get_data$status_code != 200){
    content_data <- httr::content(get_data, "text", encoding = "UTF-8")
    content_data <- jsonlite::fromJSON(content_data)
    my_message <- paste(content_data$message, collapse = "\n")
    my_message2 <- paste0('\nConsider calling retrieve_tables() to see available tables.')
    message(paste(my_message, my_message2))
    return(NULL)
  } else {
    message("Metadata collected succesfully")
  }

  content_data <- httr::content(get_data, "text", encoding = "UTF-8")

  content_data <- jsonlite::fromJSON(content_data)

  return(content_data)
}

# retrieve_data ----

#' Retrieves a specific table from Statistics Denmark
#'
#' The function retrieves a specific table from Statistics Denmark based on a
#' table ID and some parameters to the API.
#'
#' @param table_id is the id of the table you want to call. You can get table ids
#'     by calling the \link{sdk_retrieve_tables} function.
#' @param ... are parameters you need to use to specify what data you want from
#'     the API. See the data created by the \link{sdk_retrieve_metadata} function.
#'     An * indicates ALL settings in the given parameter.
#' @param base_url is the base url for the API you wish to call. Statistics Denmark
#'     can sometimes create custom API's that you can use by changing this
#'     parameter.
#' @param lang whether to return the data in english or danish.
#'
#' @return a data frame
#' @export
#'
#' @examples
#'
#' metadata <- statsDK::sdk_retrieve_metadata("FOLK1A")
#' dplyr::glimpse(metadata)
#'
#' # See the variables as a data frame
#' variables <- statsDK::sdk_get_variables(metadata)
#' dplyr::glimpse(variables)
#'
#' # Use the param and the settings columns from the variables data to set the
#' # rigth values for the API call.
#' df_en <- statsDK::sdk_retrieve_data("PRIS111", VAREGR = "000000", ENHED = "*",
#'                                 Tid = paste(paste0("2017M0", 1:8), collapse = ","))
#' dplyr::glimpse(df_en)
#'
#' df_da <- statsDK::sdk_retrieve_data("PRIS111", VAREGR = "000000", ENHED = "*",
#'                                 Tid = paste(paste0("2017M0", 1:8), collapse = ","),
#'                                 lang = "da")
#' dplyr::glimpse(df_da)

sdk_retrieve_data <- function(table_id, ..., lang = "en", base_url = "http://api.statbank.dk/v1/"){

  # Get table meta data ----
  metadata <- suppressMessages(sdk_retrieve_metadata(table_id, base_url))
  data_ids <- metadata$variables$id
  data_text <- metadata$variables$text

  # Choose baseurl based on language setting ----
  if(lang != "en"){
    base_url <- paste0(base_url, "data/",
                       table_id, "/BULK?")
  } else {
    base_url <- paste0(base_url, "data/",
                       table_id, "/BULK?lang=en")
  }

  # Params and url builder ----
  params <- list(...)

  # All params must be declared. Fill out if missing.
  missing_ids <- data_ids[!(data_ids %in% names(params))]
  if(length(missing_ids) > 0) message(paste("Autosetting missing values to * for:",
                                            paste(missing_ids, collapse = ", ")))
  missing_params <- lapply(missing_ids, function(i) "*")
  names(missing_params) <- missing_ids
  params <- c(params, missing_params)

  # Build url
  url_string <- paste0(names(params), "=", unlist(params))
  url_string <- paste(url_string, collapse = "&")

  if(lang == "en"){ url <- paste0(base_url, "&", url_string)} else {
    url <- paste0(base_url, url_string)
  }

  url <- utils::URLencode(url)

  # Get data ----
  # Notify user
  message("Getting data. This can take a while, if the data is very large.")

  get_data <- suppressWarnings(httr::GET(url))

  # Decide if error ----
  if(get_data$status_code != 200){
    content_data <- httr::content(get_data, "text", encoding = "UTF-8")
    content_data <- jsonlite::fromJSON(content_data)
    my_message <- paste(content_data$message, collapse = "\n")
    my_message2 <- paste0('\nConsider calling retrieve_metadata("', table_id,
                          '") to see available parameters.')
    message(paste(my_message, my_message2))
    return(NULL)
  } else {
    message("Data collected succesfully")
  }

  # Extract data ----
  content_data <- httr::content(get_data, "text")
  content_data <- readr::read_csv2(content_data,
                                   locale = readr::locale(
                                     decimal_mark = ",",
                                     grouping_mark = ".",
                                     tz = "CET"
                                   ))

  # Add metadata as attribute ----
  attr(content_data, "metadata") <- metadata

  # Return the data ----
  return(content_data)

}


vignettes/Get_started_with_statsDK.R
## ---- warning=FALSE, message=FALSE---------------------------------------
library(statsDK); library(dplyr); library(stringr); library(lubridate); library(ggplot2); library(tidyr)

tables <- sdk_retrieve_tables()

glimpse(tables)

## ------------------------------------------------------------------------
tables_long <- tables %>%
  unnest(variables)

## ------------------------------------------------------------------------

marriage_tables <- tables_long %>%
  filter(str_detect(tables_long$variables, "marriage"))

glimpse(marriage_tables)


## ------------------------------------------------------------------------
viedag_meta <- sdk_retrieve_metadata("VIEDAG")

glimpse(viedag_meta)

## ------------------------------------------------------------------------
variables <- sdk_get_variables(viedag_meta)

glimpse(variables)

## ------------------------------------------------------------------------
variable_overview <- variables %>% 
  group_by(param) %>%
  slice(c(1, round(n()/2), n())) %>%
  ungroup()

variable_overview

## ------------------------------------------------------------------------
VIEDAG <- sdk_retrieve_data("VIEDAG", Tid = "*", VDAG = "TOT", VIMDR = "006,012")
names(VIEDAG) <- c("time", "day", "month", "count")

## ------------------------------------------------------------------------
glimpse(VIEDAG)

## ---- fig.width=4*1.618, fig.height=4------------------------------------

VIEDAG$time <- ymd(paste0(VIEDAG$time, "-01-01"))

my_y <- VIEDAG %>%
  filter(time == max(time)) %>%
  pull(count)

ggplot(VIEDAG) +
  geom_line(aes(x = time, count, group = month)) +
  annotate("text", x = max(VIEDAG$time) %m+% months(1) , y = my_y, 
           label = c("June", "December"), hjust = 0) +
  annotate("point", x = max(VIEDAG$time), y = my_y) +
  xlim(min(VIEDAG$time), max(VIEDAG$time) %m+% years(1) ) +
  labs(y = "Total marriages for the given month", x = "Years") +
  theme_minimal()

